### ISSUE TITLE
Incrementally change the actions in the menu

### ISSUE THREAD

**Issue by marcoieni** [stevenguh/spacecode#7]
## idea

Right now the only way to customize the actions is to completely redefine `"spacecode.bindings"` in your `settings.json`.

There should be a way to add or override single key bindings without deleting all the others.

## example

If I put this in my `settings.json`, `SPC f f` should be edited, `SPC f q` should be added because it doesn't exist yet and all the other key bindings should remain as defined by default.
```
{
  "spacecode.mybindings": [
    {
      "key": "f",
      "name": "File...",
      "type": "bindings",
      "bindings": [
        {
          "key": "f",
          "name": "New Action Name",
          "type": "command",
          "command": "action.which.I.prefer.over.default"
        },
        {
          "key": "q",
          "name": "New Action Name",
          "type": "command",
          "command": "my.action.for.this.new.key.binding"
        },
        
      ]
    }
  ]
}
```

**Comment by stevenguh** [comment]
I agree that there should be a way to add, modify, and delete the default keys without copying the whole default. I need to think about what's the best way to do this.

**Comment by stevenguh** [comment]
Since the bindings are an array, which means they are positional, and that makes the override more complicated. I am trying the following override config and hopefully that will accomplish the full spectrum of edits on the existing config

```jsonc
{
  "spacecode.overrides": [
    // Replace `<SPC> f` completely with the specified bindings and move it to position 0 if possible
    {
      "keys": "f",
      "name": "File...",
      "type": "bindings",
      "position": 0,
      "bindings": [
        {
          "key": "f",
          "name": "New Action Name",
          "type": "command",
          "command": "action.which.I.prefer.over.default"
        },
        {
          "key": "q",
          "name": "New Action Name",
          "type": "command",
          "command": "my.action.for.this.new.key.binding"
        },
      ]
    },
    // Replace the `<SPC> g s` if exist.
    // Append if it doesn't exist
    {
      "keys": "g.s"
      "name": "Go to",
      "type": "command",
      "command": "workbench.action.gotoLine",   
    },
    // Remove anything that bind to `<SPC> s t`
    // Negative position denote removal
    {
      "keys": "s.t"
      "position": -1
    },
    // Move `<SPC> g` to at index 2 if possible
    {
      "keys": "g"
      "position": 2
    },
  ]
}
```

**Comment by marcoieni** [comment]
In this way the user is responsible for keeping an alphabetical order? We can do better. 

Maybe we could:
- search for the right insert position by ourselves
- keep the data in a map where the key is the key of the keyboard. In this way when the user is typing the access is very fast, too.

Anyway I am not familiar with your codebase or with vscode extensions, therefore this may be just a rambling speech.

**Comment by marcoieni** [comment]
Oh, and I think that the dot in "key": "g.s" is confusing, I would delete it.

**Comment by stevenguh** [comment]
The user is responsible for keeping order cause I am not sure if alphabetical order works for everyone. That's why the optional `position` exists to enforce override position. If `position` is not provided, the override will be appended the if that key doesn't exist, and replaced in the same position if that key already exists. If `position` is provided, the override will be insert at that position the if that key doesn't exist, and replaced and insert at that position if that key already exists.

> search for the right insert position by ourselves

If the key doesn't exist, it would be tricky to search the "right" insert. If appending is a good enough default, then I think the schema I had can facilitate when the user omit the `position` key.

> keep the data in a map where the key is the key of the keyboard. In this way when the user is typing the access is very fast, too.

I didn't use map in the existing data structure to provide customization of ordering.

> Oh, and I think that the dot in "key": "g.s" is confusing, I would delete it.

The dot notation is there to surgically replace element with that key combo.
For example, the following json will surgical replace `<SPC> g s` to go to line while keeping all other elements in `<SPC> g`
```json
{
      "keys": "g.s",
      "name": "Go to",
      "type": "command",
      "command": "workbench.action.gotoLine",   
}
```

This following json will replaced the whole `<SPC> g` menu with a single binding.
```json
{
      "keys": "g",
      "name": "Go...",
      "type": "bindings",
      "bindings": [{
          "key": "s",
          "name": "Go to",
          "type": "command",
          "command": "workbench.action.gotoLine",   
        }]
}
```

The choice of the dot is deliberate so that people can replace/insert/delete the only key that they need or the whole bindings. Think of the overrides as a list of bindings that needs to be surgically replaced.

**Comment by marcoieni** [comment]
Oh! Sorry, I didn't understand that `position` is an optional field. 😅

> If position is not provided, the override will be appended the if that key doesn't exist, and replaced in the same position if that key already exists. If position is provided, the override will be insert at that position the if that key doesn't exist, and replaced and insert at that position if that key already exists.

This mechanism seems reasonable to me. :+1: 

> If appending is a good enough default, then I think the schema I had can facilitate when the user omit the position key.

If "append by default" helps you at the moment just do it. This behavior could be modified later, maybe by another contributor.

> The dot notation is there to surgically replace element with that key combo.

Ok, in my mind the last example you provided just adds the `SPC g s` command (you can see it from the example I provided). But now that I think about it, providing `SPC g s` in a single line is more concise, so maybe it is better.
Have you considered `"keys": "gs",` instead of `"keys": "g.s",`

**Comment by stevenguh** [comment]
Sorry for not explaining clearly. Thank you for going through the issue with me :)

> Have you considered `"keys": "gs"`, instead of `"keys": "g.s",`

The use of keys without dots will assume each key is only single character, however, the current menu doesn't impose that as a limitation. Also I think the dot notation is clearer to indicate different level of nesting, as well familiarities of traversing object in javascript

**Comment by marcoieni** [comment]
> The use of keys without dots will assume each key is only single character, however, the current menu doesn't impose that as a limitation.

Wow, you are covering a lot of edge cases. If some people like to type more it's ok.

> Also I think the dot notation is clearer to indicate different level of nesting

I find that without the dot is clearer.

Anyway, I expressed all my concerns about this issue, final decisions are up to you 😁

> Thank you for going through the issue with me

Thank you for working on this project!

**Comment by stevenguh** [comment]
Thank you for the feedback. I hope this extension is helpful to people :) I just pushed the changes and released `v0.3.2`
